### ISSUE TITLE
How to translate index-assignment

### ISSUE THREAD

**Issue by vini-fda** [LaurentMazare/tch-rs#770]
I'm reading a [PyTorch tutorial on RL](https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html) and would like to know how to translate this to tch-rs code:

```python
# Compute V(s_{t+1}) for all next states.
# Expected values of actions for non_final_next_states are computed based
# on the "older" target_net; selecting their best reward with max(1)[0].
# This is merged based on the mask, such that we'll have either the expected
# state value or 0 in case the state was final.
next_state_values = torch.zeros(BATCH_SIZE, device=device)
with torch.no_grad():
    next_state_values[non_final_mask] = target_net(non_final_next_states).max(1)[0]
```

Specifically, how do I index with a mask and assign to that indexed tensor?

**Comment by LaurentMazare** [comment]
You can check the [tch::index documentation](https://docs.rs/tch/latest/tch/index/index.html) for various ways of indexing tensors, and then use `copy_` on the result.
However note that indexing is likely to create a fresh tensor when using complex indexes such as a mask and so the assignment would not propagate to the original values. At that point it really depends on the shape of your mask, if it's simple enough, you may want to use a combination of `narrow` and `copy`, you can see some example of this in the [tch reinforcement learning examples](https://github.com/LaurentMazare/tch-rs/blob/main/examples/reinforcement-learning/ppo.rs#L59-L63) and I would suggest looking a bit at these examples as it's likely to have content relevant to implementing the PyTorch tutorial.

**Comment by vini-fda** [comment]
Ok! I've looked in these examples, but I'm not sure I found what I was specifically looking for: overwriting into a Tensor that is indexed with a Tensor mask (I'll update my question to be more specific). One promising solution that worked for me for now was using `index_put_`:

```rust
tch::no_grad(|| {
    let expected_values = self
        .target_net
        .forward(&non_final_next_states)
        .max_dim(1, false)
        .0;
    let _ = next_state_values.index_put_(&[Some(non_final_mask)], &expected_values, false);
});
```
